### ISSUE TITLE
Another bug in RAJA function that only happens on marianas

### ISSUE THREAD

**Issue by nychiang** [LLNL/hiop#255]
Branch `fr-mds-dev` passes all the unit test on newell and ascent, but gets problems again on marianas.
This is not about the `long long` --> `int` conversion bug, and I created branch `raja-marianas-bug-fix` to reproduce and narrow down the bug.

in short, I received CUDA assertion error [on this line ](https://github.com/LLNL/hiop/blob/96f793f6311c0950bc9e360e4a78f8c3d8a5942e/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp#L1138).
Note that I add this fake line to show the problem is not caused by accessing array out of bounds. 
'this' matrix is initialized [here](https://github.com/LLNL/hiop/blob/96f793f6311c0950bc9e360e4a78f8c3d8a5942e/tests/testMatrixSparse.cpp#L246), given a space to save 105 non-zeros.

The error message is " CUDAassert: an illegal memory access was encountered /qfs/projects/exasgd/spack-install/linux-centos7-broadwell/gcc-7.3.0/raja/0.13.0-pwmbk/include/RAJA/policy/cuda/MemUtils_CUDA.hpp 183"

I am guessing here we have another implicit type conversion that fails CUDA.

PS: FYI, the RAJA::Index_type type is an alias to the C++ type std::ptrdiff_t, and std::ptrdiff_t is the signed integer type of the result of subtracting two pointers.

@pelesh @cnpetra @ashermancinelli

**Comment by nychiang** [comment]
From the PNNL pipeline results, we can see the unit test can reach [line1155](https://github.com/LLNL/hiop/blob/96f793f6311c0950bc9e360e4a78f8c3d8a5942e/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp#L1155) and print some information on screen, after we remove [ line 1138](https://github.com/LLNL/hiop/blob/96f793f6311c0950bc9e360e4a78f8c3d8a5942e/src/LinAlg/hiopMatrixRajaSparseTriplet.cpp#L1138).

I probably have run 50+ tests in PNNL pipeline. Please ignore the old ones and compare the latest two results, i.e., the tests for commit 268281aa (without line 1138) and  96f793f6.

**Comment by cnpetra** [comment]
I think there is an out-of-bounds write on the device that messes up the `iRow_` pointer and causes `iRow_[0] = 0;` to fail. Is there any way to "valgrind" on the device?

**Comment by nychiang** [comment]
I tried to use valgrind on ascent but it didn't throw any waring/error in this function.

On marianas, I tried to add the following code to test if `iRow` and `values` can be modified by UM
```
  copyFromDev();            // copy the sparse triplet from device
  iRow_host_[0]=1;           // change sparse triplet on host
  values_host_[0]=2.0;    // change sparse triplet on host
  copyToDev();                 // copy the sparse triplet to device
  setToConstant(5.0);      // change nonzero value on device via UM
  copyFromDev();            // copy the sparse triplet from device
  std::cout<<"1st val on host: " << values_host_[0]<<std::endl;  // this line prints out `5`
```
and change `RAJA::Index_type` to `int`, but still got problem when accessing `iRow_[0]`

@ashermancinelli May I have access to marianas? In the meantime, I wonder if you could help me to run valgrind on marianas and show me the log file?

**Comment by ashermancinelli** [comment]
Access to marianas is more of an institutional issue, I can't do anything there...  I can certainly build/run whatever you need there though. Is there a particular branch/binary you'd like me to use valgrind on?

**Comment by nychiang** [comment]
@ashermancinelli Can you use valgrind on branch `raja-marianas-bug-fix`? Thanks!

**Comment by ashermancinelli** [comment]
Any particular executable? I'll assume unit tests and example 4?

**Comment by nychiang** [comment]
Unit test `tests/testMatrixSparse` is enough. Thanks!

**Comment by cnpetra** [comment]
I think this has been fixed by PR#239, please confirm @nychiang

**Comment by nychiang** [comment]
Yes. It is fixed after I remove `this` pointer in the RAJA code in PR#239. Close this issue.
